File: trackbuilder.py
import math
import cairo

NUM_CL = 6
NUM_CR = 6
NUM_S = 4

class TrackBuilder(object):

	def __init__(self):
		self.r = 40.0
		self.WIDTH = 400
		self.HEIGHT = 400
		self.startx = self.WIDTH/2
		self.starty = self.HEIGHT/2
		self.currentx = 0.0
		self.currenty = 0.0
		self.currdeg = 270
		self.arcx = math.cos(math.radians(45)) * self.r
		self.arcy = math.sin(math.radians(45)) * self.r
		self.linewidth = 5
		self.path = None
		self.surface = None
		self.ctx = None

	def reset_ctx(self, path):
		self.path = path
		self.surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, int(self.WIDTH), int(self.HEIGHT))
		self.ctx = cairo.Context(self.surface)
		self.ctx.move_to(self.startx, self.starty)
		self.currentx, self.currenty = self.ctx.get_current_point()

	@staticmethod
	def is_valid_path(path):
		t1 = True
		# check for valid use of 's'
		if path.count('s') >= 3:
			l2 = ''.join(path * 2)
			if l2.find('ss') == -1:
				t1 = False

		# the curves are reversible so we need to check that number <= cl + cr
		t2 = (path.count('cl') + path.count('cr') <= NUM_CL + NUM_CR) and path.count('s') <= NUM_S
		return t1 and t2

	def add(self, t):
		if t == 'cr':
			# new sub path required so arc doesn't close
			self.ctx.new_sub_path()
			if self.currdeg == 0:
				self.ctx.arc(self.currentx, self.currenty+self.r, self.r, math.radians(self.currdeg-90), math.radians(self.currdeg+45-90))
			elif self.currdeg == 45:
				self.ctx.arc(self.currentx - self.arcx, self.currenty + self.arcy, self.r, math.radians(self.currdeg-90), math.radians(self.currdeg+45-90))
			elif self.currdeg == 90:
				self.ctx.arc(self.currentx - self.r, self.currenty, self.r, math.radians(self.currdeg-90), math.radians(self.currdeg+45-90))
			elif self.currdeg == 135:
				self.ctx.arc(self.currentx - self.arcx, self.currenty - self.arcy, self.r, math.radians(self.currdeg-90), math.radians(self.currdeg+45-90))
			elif self.currdeg == 180:
				self.ctx.arc(self.currentx, self.currenty-self.r, self.r, math.radians(self.currdeg-90), math.radians(self.currdeg+45-90))
			elif self.currdeg == 225:
				self.ctx.arc(self.currentx+self.arcx, self.currenty-self.arcy, self.r, math.radians(self.currdeg-90), math.radians(self.currdeg+45-90))
			elif self.currdeg == 270:
				self.ctx.arc(self.currentx+self.r, self.currenty, self.r, math.radians(self.currdeg-90), math.radians(self.currdeg+45-90))
			elif self.currdeg == 315:
				self.ctx.arc(self.currentx+self.arcx, self.currenty+self.arcy, self.r, math.radians(self.currdeg-90), math.radians(self.currdeg+45-90))

			self.currdeg = self.currdeg + 45
			if self.currdeg == 360:
				self.currdeg = 0 

		if t == 'cl':
			# new sub path required so arc doesn't close
			self.ctx.new_sub_path()
			if self.currdeg == 0:
				self.ctx.arc_negative(self.currentx, self.currenty-self.r, self.r, math.radians(self.currdeg+90), math.radians(self.currdeg-45+90))
			elif self.currdeg == 45:
				self.ctx.arc_negative(self.currentx+self.arcx, self.currenty-self.arcy, self.r, math.radians(self.currdeg+90), math.radians(self.currdeg-45+90))
			elif self.currdeg == 90:
				self.ctx.arc_negative(self.currentx+self.r, self.currenty, self.r, math.radians(self.currdeg+90), math.radians(self.currdeg-45+90))
			elif self.currdeg == 135:
				self.ctx.arc_negative(self.currentx+self.arcx, self.currenty+self.arcy, self.r, math.radians(self.currdeg+90), math.radians(self.currdeg-45+90))
			elif self.currdeg == 180:
				self.ctx.arc_negative(self.currentx, self.currenty+self.r, self.r, math.radians(self.currdeg+90), math.radians(self.currdeg-45+90))
			elif self.currdeg == 225:
				self.ctx.arc_negative(self.currentx-self.arcx, self.currenty+self.arcy, self.r, math.radians(self.currdeg+90), math.radians(self.currdeg-45+90))
			elif self.currdeg == 270:
				self.ctx.arc_negative(self.currentx-self.r, self.currenty, self.r, math.radians(self.currdeg+90), math.radians(self.currdeg-45+90))
			elif self.currdeg == 315:
				self.ctx.arc_negative(self.currentx-self.arcx, self.currenty-self.arcy, self.r, math.radians(self.currdeg+90), math.radians(self.currdeg-45+90))

			if self.currdeg == 0:
				self.currdeg = 360
			self.currdeg = self.currdeg - 45

		elif t == 's':
			if self.currdeg == 0:
				self.ctx.line_to(self.currentx+self.r, self.currenty)
			elif self.currdeg == 45:
				self.ctx.line_to(self.currentx+math.sqrt(math.pow(self.r, 2)/2), self.currenty+math.sqrt(math.pow(self.r,2)/2))
			elif self.currdeg == 90:
				self.ctx.line_to(self.currentx, self.currenty+self.r)
			elif self.currdeg == 135:
				self.ctx.line_to(self.currentx-math.sqrt(math.pow(self.r, 2)/2), self.currenty+math.sqrt(math.pow(self.r,2)/2))
			elif self.currdeg == 180:
				self.ctx.line_to(self.currentx-self.r, self.currenty)
			elif self.currdeg == 225:
				self.ctx.line_to(self.currentx-math.sqrt(math.pow(self.r, 2)/2), self.currenty-math.sqrt(math.pow(self.r,2)/2))
			elif self.currdeg == 270:
				self.ctx.line_to(self.currentx, self.currenty-self.r)
			elif self.currdeg == 315:
				self.ctx.line_to(self.currentx+math.sqrt(math.pow(self.r, 2)/2), self.currenty-math.sqrt(math.pow(self.r,2)/2))

		self.currentx, self.currenty = self.ctx.get_current_point()

	def build(self, pieces):
		start_point = self.currentx, self.currenty
		start_deg = self.currdeg
		for p in pieces:
			self.add(p)
		end_point = self.currentx, self.currenty
		end_deg = self.currdeg
		return start_point == end_point and start_deg == end_deg

	def init_image(self):

		# draw a background
		(x, y, z) = (1, 1, 1)
		self.ctx.set_source_rgb(x, y, z)
		self.ctx.rectangle(0, 0, self.WIDTH, self.HEIGHT)
		self.ctx.fill()

		# draw a border
		(x, y, z) = (165/255, 165/255, 165/255)
		self.ctx.set_source_rgb(x, y, z)
		self.ctx.rectangle(0.0, 0.0, self.WIDTH, self.HEIGHT)
		self.ctx.stroke()

		# draw the path of the image
		self.ctx.move_to(5, 10)
		self.ctx.show_text(str(self.path))

		# draw point to represent starting point
		self.ctx.move_to(self.startx, self.starty)
		self.ctx.arc(self.startx, self.starty, 5, 0, 2*math.pi)
		self.ctx.set_source_rgb(0.3, 0.4, 0.6)
		self.ctx.fill()

		self.ctx.move_to(self.startx, self.starty)

	def create_image(self, imageid):

		self.reset_ctx(self.path)
		self.init_image()
		self.build(self.path)
		self.ctx.stroke()

		# draw the id of the image
		self.ctx.move_to(5, 25)
		self.ctx.show_text(imageid)
		self.surface.write_to_png(imageid + ".png")

File: example/TrackBuilder.cpp


#include <cmath>
#include <string>
#include "cairo.h"

#define degtorad(angleDegrees) (angleDegrees * M_PI / 180.0)
#define radtodeg(angleRadians) (angleRadians * 180.0 / M_PI)

class TrackBuilder {

public:

	float getCurrentX() { return currentx; }
	float getCurrentY() { return  currenty; }
	int getCurrentDegree() { return  currdeg; }

/*
	void createImage (int imageid) {
		init(path)
		initimage()
		buildit(path)
		ctx.stroke()

		# draw the id of the image
		ctx.move_to(5, 25)
		ctx.show_text(imageid)
		surface.write_to_png (imageid + ".png")
*/

protected:

	void init(std::string path) {
		//path = path;
		//surface = cairo_image_surface_create (FORMAT_ARGB32, WIDTH, HEIGHT);
		//ctx = cairo.Context (surface)
		//ctx.move_to (startx, starty)
		//currentx = ctx.get_current_point()
	}

private:
	std::string path;
	float r = 40.0;
	const int WIDTH = 400;
	const int HEIGHT = 400;
	float startx = WIDTH/2;
	float starty = HEIGHT/2;
	float currentx = 0.0;
	float currenty = 0.0;
	int currdeg = 270;
	float arcx = cos(degtorad(45)) * r;
	float arcy = sin(degtorad(45)) * r;
	int linewidth = 5;
	cairo_surface_t *surface;

};


/*

	def initimage(:

		# draw a background
		(x, y, z) = rgb(255,255,255)
		ctx.set_source_rgb(x, y, z)
		ctx.rectangle(0, 0, WIDTH, HEIGHT)
		ctx.fill()

		# draw a border 
		(x, y, z) = rgb(165, 165, 165)
		ctx.set_source_rgb(x, y, z)
		ctx.rectangle(0.0, 0.0, WIDTH, HEIGHT);
		ctx.stroke();

		# draw the path of the image
		ctx.move_to(5, 10)
		ctx.show_text(str(path))

		# draw point to represent starting point
		ctx.move_to (startx, starty)
		ctx.arc(startx, starty, 5, 0, 2*math.pi)
		ctx.set_source_rgb(0.3, 0.4, 0.6)
		ctx.fill()

		ctx.move_to (startx, starty)

	def rgb(x,y,z):
		return (x/255, y/255, z/255)

	def buildit( pieces):
		for p in pieces:
			add(p)

	def add(t):
		if t == 'cr':
			# new sub path required so arc doesn't close
			ctx.new_sub_path()
			if currdeg == 0:
				ctx.arc (currentx, currenty+r, r, math.radians(currdeg-90), math.radians(currdeg+45-90))
			elif currdeg == 45:
				ctx.arc (currentx - arcx, currenty + arcy, r, math.radians(currdeg-90), math.radians(currdeg+45-90))
			elif currdeg == 90:
				ctx.arc (currentx - r, currenty, r, math.radians(currdeg-90), math.radians(currdeg+45-90))
			elif currdeg == 135:
				ctx.arc (currentx -arcx, currenty -arcy, r, math.radians(currdeg-90), math.radians(currdeg+45-90))
			elif currdeg == 180:
				ctx.arc (currentx, currenty-r, r, math.radians(currdeg-90), math.radians(currdeg+45-90))
			elif currdeg == 225:
				ctx.arc (currentx+arcx, currenty-arcy, r, math.radians(currdeg-90), math.radians(currdeg+45-90))
			elif currdeg == 270:
				ctx.arc (currentx+r, currenty, r, math.radians(currdeg-90), math.radians(currdeg+45-90))
			elif currdeg == 315:
				ctx.arc (currentx+arcx, currenty+arcy, r, math.radians(currdeg-90), math.radians(currdeg+45-90))

			currdeg = currdeg + 45
			if currdeg == 360:
				currdeg = 0 

		if t == 'cl':
			# new sub path required so arc doesn't close
			ctx.new_sub_path()
			if currdeg == 0:
				ctx.arc_negative (currentx, currenty-r, r, math.radians(currdeg+90), math.radians(currdeg-45+90))
			elif currdeg == 45:
				ctx.arc_negative (currentx+arcx, currenty-arcy, r, math.radians(currdeg+90), math.radians(currdeg-45+90))
			elif currdeg == 90:
				ctx.arc_negative (currentx+r, currenty, r, math.radians(currdeg+90), math.radians(currdeg-45+90))
			elif currdeg == 135:
				ctx.arc_negative (currentx+arcx, currenty+arcy, r, math.radians(currdeg+90), math.radians(currdeg-45+90))
			elif currdeg == 180:
				ctx.arc_negative (currentx, currenty+r, r, math.radians(currdeg+90), math.radians(currdeg-45+90))
			elif currdeg == 225:
				ctx.arc_negative (currentx-arcx, currenty+arcy, r, math.radians(currdeg+90), math.radians(currdeg-45+90))
			elif currdeg == 270:
				ctx.arc_negative (currentx-r, currenty, r, math.radians(currdeg+90), math.radians(currdeg-45+90))
			elif currdeg == 315:
				ctx.arc_negative (currentx-arcx, currenty-arcy, r, math.radians(currdeg+90), math.radians(currdeg-45+90))

			if currdeg == 0:
				currdeg = 360
			currdeg = currdeg - 45

		elif t == 's':
			if currdeg == 0:
				ctx.line_to (currentx+r, currenty)
			elif currdeg == 45:
				ctx.line_to (currentx+math.sqrt(math.pow(r,2)/2), currenty+math.sqrt(math.pow(r,2)/2))
			elif currdeg == 90:
				ctx.line_to (currentx, currenty+r)
			elif currdeg == 135:
				ctx.line_to (currentx-math.sqrt(math.pow(r,2)/2), currenty+math.sqrt(math.pow(r,2)/2))
			elif currdeg == 180:
				ctx.line_to (currentx-r, currenty)
			elif currdeg == 225:
				ctx.line_to (currentx-math.sqrt(math.pow(r,2)/2), currenty-math.sqrt(math.pow(r,2)/2))
			elif currdeg == 270:
				ctx.line_to (currentx, currenty-r)
			elif currdeg == 315:
				ctx.line_to (currentx+math.sqrt(math.pow(r,2)/2), currenty-math.sqrt(math.pow(r,2)/2))

		currentx, currenty = ctx.get_current_point()

*/
